src/tools/miri/src/bin/miri.rs

```diff
@@ -244,6 +269,10 @@ struct MiriBeRustCompilerCalls {
 impl rustc_driver::Callbacks for MiriBeRustCompilerCalls {
     #[allow(rustc::potential_query_instability)] // rustc_codegen_ssa (where this code is copied from) also allows this lint
     fn config(&mut self, config: &mut Config) {
+        if self.target_crate {
```

Comment:
Curious that this needs a slightly different condition than the next if block... any idea why?

Reply:
Cargo runs `rustc --print file-names --crate-type rlib --crate-type dylib ...` and then expects for every crate type either the filename or a warning that the crate type isn't supported. In the latter case cargo will not show the warning, but will record the fact that the crate type is not supported and for example error out if a crate only gets compiled with unsupported crate types. If we filter unsupported `--crate-type`, the warning will never be emitted and cargo will report an error that it failed to parse the output of rustc. This `crate_types.retain()` call is to prevent the unsupported crate type warning when compiling for real, in which case `--print` is not passed.

Reply:
I see... I think I understand. I added a commit rearranging the code a bit and adding comments. Does that look right?